### PR TITLE
Generate assets from a semantic convention registry and a set of jinja2 templates

### PR DESCRIPTION
**Pull request by lquerel**
This PR sets up the framework for generating code or documentation for semantic conventions. Integration into the command line is not part of this PR, but a unit test in `weaver_forge` demonstrates and tests the generation possibilities.

Documentation on using the template engine is available in the `docs/template-engine.md` directory.

The template engine is based on the `minijinja` crate. This template engine is largely compatible with the Jinja2 engine (Python ecosystem).

### CODE REVIEW

File: docs/template-engine.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
+# Template Engine 
+
+OTel Weaver is capable of generating documentation or code from a semantic
```
Comment (jsuereth) [inline docs/template-engine.md:3]: Before we spend more time going down this path, we should validate this is desired from the rest of SEmconv maintainers.


There are several types of codegen I think we'll want going forward:


1. We have human-written instructions for semantic conventions where-in we *embed* rendered markdown outlining the state of the semconv registry, via `<-- semconv( render command) -->`.  I'd prefer if we worked towards feature parity on this behavior first.
2. We have basic rendering of the "registry itself" as a generic "{language}doc" like output.  While I don't think that's exactly what you had in mind in this PR, it's the second thing I'd like to see.
3. We will need to generate "{langauge}doc" like output for "application" instrumentation.  I believe this is what you're creating here, and I really like the direction, but think we should first make sure 1+2 are solved.
4. We will need to generate code.

I actually think we should keep these four use cases as *separate* things for weaver.

TL;DR; I think this currently makes another good demo, but we need to start socializing it so it can go further.   Additionally, let's see if we can replicate some key use cases of semconv generation today.

**Review by jsuereth — APPROVED** [review]
Approving but a few caveats for the future:


- Next time I'd prefer a separate PR that updates the resolver for changes in stability, e.g.  That'd be a lot simpler to review and merge if not bundled with the template change.
- I think you and I may need to sit down and figure out priorities for this project.  You can see my comment on templating and feature set.  I think we *both* agree we need to get to a "resolved schema" file format ASAP and propose this in the specification, but we may disagree on how to get there.  My current thinking is showcase that weaver is 1:1 compatible w/ buld-tools for just the semconv-repo itself *and* outline the resolved schema it produces along with an accompanying specification.

In any case, nice work with this!  Exciting to see that the template language isn't a huge jump away from JINJA.

File: crates/weaver_forge/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,762 @@
...
+        })
+    }
+
+    // ToDo Refactor InternalError
+    // ToDo Use compound error
+
+    /// Generate assets from a semantic convention registry.
+    pub fn generate_registry(
+        &self,
+        log: impl Logger + Clone + Sync,
```
[the remark this answers is not in the document]
Comment (lquerel) [reply crates/weaver_forge/src/lib.rs:257]: Good idea. I will make this change in a future PR.

File: crates/weaver_forge/src/registry.rs [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        registry: &Registry,
+        catalog: &Catalog,
+    ) -> Result<Self, Error> {
+        let mut errors = Vec::new();
+
+        let groups = registry
+            .groups
+            .iter()
+            .map(|group| {
+                let id = group.id.clone();
```
[the remark this answers is not in the document]
Comment (lquerel) [reply crates/weaver_forge/src/registry.rs:117]: I don't think it will make a big difference in this case, but in principle, it makes sense to link the lifetimes of these two structures.
